### ISSUE TITLE
Invalid return type inference in function with multiple return points 

### ISSUE THREAD

**Issue by smallSwed** [microsoft/pyright#4654]
**Describe the bug**
Invalid type inference for function return types if multiple type of dictionaries are returned and one of them is initiated with an empty `dict`. Pyright only considers one of the return points for inferring the return type, ignoring the empty dictionary type which is `dict[Unknown, Unknown]`.

**To Reproduce**
Steps to reproduce the behavior: open code in vscode with pyright extension 
```python
def foo(error: bool): # inferred type: (function) def foo(error: bool) -> dict[str, dict[str, str]]
    result = {}
    if error:
        return {"error": {"message": "some message"}}
    result["list"] = []
    return result

def bar(error: bool): # inferred type: (function) def foo(error: bool) -> dict[str, dict[str, str]]
    result = {}
    if not error:
        result["list"] = []
        return result
    return {"error": {"message": "some message"}}

result = foo(False) # same for bar(...)
result["list"].append("something") # error: Cannot access member "append" for type "dict[str, str]"
```

**Expected behavior**
Instead of `dict[str, dict[str, str]]` should infer `dict[str, dict[str, str]]|dict[Unknown, Unknown]` to avoid misleading error messages at the calling point of the function.  

**VS Code extension**
Pyright v1.1.294
**VS Code** 
Version: 1.75.1
Commit: 441438abd1ac652551dbe4d408dfcec8a499b8bf
Date: 2023-02-08T21:35:30.018Z
Electron: 19.1.9
Chromium: 102.0.5005.194
Node.js: 16.14.2
V8: 10.2.154.23-electron.0
OS: Linux x64 5.4.0-139-generic snap
Sandboxed: No

**Comment by erictraut** [comment]
I agree that there's a bug here. The correct inferred return type should be `dict[str, dict[str, str]] | dict[Unknown, Unknown]` as you indicated. However, that won't eliminate the error in your code. You still have a type violation. The type of `result["list"]` is `dict[str, str] | Unknown]`, so you cannot call `append` with a value of `"something"`.

The return type inference logic will be fixed in the next release.

**Comment by Sammymullern** [comment]
The error is occurring because the result dictionary returned by the `foo` and` bar` functions does not initially have a list key. When foo is called with error = False, it returns a dictionary with a list key, but the list key's value is an empty list. When you try to append an item to the list using result["list"].append("something"), an error occurs because the list key is not yet present in the dictionary returned by foo, and therefore has no append method.
To fix this error, you can modify the foo and bar functions to always return a dictionary with a list key, even if the value of the list key is an empty list. Here's an updated version of the functions:
def foo(error: bool) -> dict[str, dict[str, str]]:
    result = {"list": []}
    if error:
        return {"error": {"message": "some message"}}
    return result

def bar(error: bool) -> dict[str, dict[str, str]]:
    result = {"list": []}
    if not error:
        return result
    return {"error": {"message": "some message"}}

Now, when you call foo(False) or bar(False), you will get a dictionary with a list key that you can append items to without encountering an error.Now, when you call foo(False) or bar(False), you will get a dictionary with a list key that you can append items to without encountering an error.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.295, which I just published. It will also be included in a future release of pylance.
